### DISCUSSION TITLE
Hi Cebra team, does this model accept multiple discrete variables?

### DISCUSSION THREAD

**Discussion by XuanMabard** [AdaptiveMotorControlLab/CEBRA#127] [Q&A] [answered]
our dataset has behaviors code and position on the screen(upper-left, lower_left) as discrete variables. we are wondering if the model is capable of consume multiple discrete variables as input along with neural data?

**Comment by stes** [comment, marked as answer]
Hi @XuanMabard , thanks for asking. Yes, you can feed in discrete variables to the model. I would still recommend to cast everything to a `float` before feeding it into the model.

**Reply by XuanMabard** [reply]
thank you for your response! sorry i was't clear about my question. So far, we have multi-dimensional discrete label we want to feed this model, and we would like to consider them independently. (for example the input of the discrete label might looks like [(0,1), (1,2), (3,0), ...] where each time bin is associated with 2 variables.). Would it be possible to feed the model this way? i noticed that in the API doc, it says the Y input of the model.fit() only takes 1d array.

**Reply by stes** [reply]
@XuanMabard , thanks for clarifying. That changes the answer a bit. If you want to pass the indices as a *label* (vs. as an input to the model), then you need to flatten the data into a 1d array first.

Could you clarify if this is what you mean? This changes the outcome substantially: You could either (1) feed the discrete labels together with the neural data as an *input*, then this would be used to map into the embedding space, or (2) embed the neural data only with the discrete label information you provide to the `fit()` function.

If your question is about case (2), then you first need to flatten your 2dim label array. This can be done by remapping each individual instances to a new symbol, and then feeding that sequence in. So in your example, `[(0,1), (1,2), (3,0), (0,1),...]` could be mapped to `[1, 2, 2, 1]`, if you want to consider all unique combinations of label pairs.

Does this help? Happy to discuss further.

**Reply by XuanMabard** [reply]
Thank you so much for your response, and it's really helpful for us! we are considering feed the Cebra model with: (1) discrete labels (2) neural data as training data to feed the model. for the discrete labels input, we want to include: (1) area ID on the screen which monkey draw shapes on (2) shape id which represent the shapes monkeys are drawing on the screen. so the data for the discrete labels will looks like [(square, upper_left), (circle, upper_right), (zigzag, lower_left), (square, upper_left)........](we will convert it to integer before train it). from what i understand, the solution is reduce the 2d array to 1d, using unique combination as id. so the discrete label input will looks like[1,2,3,1....]. if i understand it correctly. thank you, again, for your speedy response!

**Reply by XuanMabard** [reply]
Hi Cebra team,
If we are using the solution one you mentioned above, we have two questions: 1) whether this make any sense that if, in our case,  the discrete labels are categorical variables with no inherent ordering. 2)would this reduce the importance of the categorical data, because it's lower dimensionality compare to the neural data.

 now if we implement option two, this might not be the best for our goals, because our goal is to identify latent variables that represent each of the categorical variables independent to the others? also, Is the requirement of having only one discrete variable a fundamental limitation of the algorithm?

We were thinking one possibility, is to train separate model for each discrete variables, do you think this make sense?

Thank you so much for your follow-up answers!

**Reply by MMathisLab** [reply]
You can train independent models to compare the goodness of fit, but you can also train models with >1 discrete variable (correct me if I'm wrong @stes )

**Reply by XuanMabard** [reply]
Hi cebra team,
I'm really interested in how many discrete variables the model can take, i would love to hear more from you. 
thank you so much!

**Reply by vince-jq-sun** [reply]
I have the same question. It seems that recode discrete 2D behavioral variables as 1D changes the meaning of the 2D variables. Is there any way to indeed feed the training with a 2D discrete variable? 

My current solution is to dummy code each value in the 2D discrete variables as continuous discrete variables of higher dimension?.For example, if my discrete variables have two columns; each column has two values, can I re-code them as a 4D continuous vairable (each columns represent the presence of one value of one variable using 1, and absence of that value as 0). 
Is such approach appropriate? Is there a direct way to input a 2D discrete variable without flattening it?

**Reply by MMathisLab** [reply]
as stated, you can test the code here: https://github.com/AdaptiveMotorControlLab/CEBRA/pull/135
